src/main/scala/beam/agentsim/infrastructure/parking/ParkingZoneSearch.scala

```diff
@@ -137,114 +143,181 @@ object ParkingZoneSearch {
     // find zones
     @tailrec
     def _search(
-      thisInnerRadius: Double,
-      thisOuterRadius: Double,
+      searchMode: SearchMode[GEO],
       parkingZoneIdsSeen: List[Id[ParkingZoneId]] = List.empty,
       parkingZoneIdsSampled: List[(Id[ParkingZoneId], Option[ChargingPointType], ParkingType, Double)] = List.empty,
       iterations: Int = 1
     ): Option[ParkingZoneSearchResult[GEO]] = {
-      if (thisInnerRadius > config.searchMaxRadius) None
-      else {
+      // a lookup of the (next) search ring for TAZs
+      searchMode.lookupParkingZones(params.zoneQuadTree) match {
+        case Some(theseZones) =>
+          // ParkingZones as as ParkingAlternatives
+          val alternatives: List[ParkingSearchAlternative[GEO]] = {
+            for {
+              zone                <- theseZones
+              parkingTypesSubtree <- params.zoneSearchTree.get(zone.getId).toList
+              parkingType         <- params.parkingTypes
+              parkingZoneIds      <- parkingTypesSubtree.get(parkingType).toList
+              parkingZoneId       <- parkingZoneIds
+              parkingZone         <- ParkingZone.getParkingZone(params.parkingZones, parkingZoneId)
+              if parkingZoneFilterFunction(parkingZone)
+            } yield {
+              // wrap ParkingZone in a ParkingAlternative
+              val stallLocation: Coord = parkingZoneLocSamplingFunction(parkingZone)
+              val stallPriceInDollars: Double =
+                parkingZone.pricingModel match {
+                  case None => 0
+                  case Some(pricingModel) if params.searchMode == ParkingSearchMode.EnRoute =>
+                    PricingModel.evaluateParkingTicket(pricingModel, config.enrouteDuration.toInt)
+                  case Some(pricingModel) =>
+                    PricingModel.evaluateParkingTicket(pricingModel, params.parkingDuration.toInt)
+                }
+              val parkingAlternative: ParkingAlternative[GEO] =
+                ParkingAlternative(zone, parkingZone.parkingType, parkingZone, stallLocation, stallPriceInDollars)
+              val parkingAlternativeUtility: Map[ParkingMNL.Parameters, Double] =
+                parkingZoneMNLParamsFunction(parkingAlternative)
+              ParkingSearchAlternative(
+                parkingAlternative,
+                parkingAlternativeUtility
+              )
+            }
+          }
 
-        // a lookup of the (next) search ring for TAZs
-        val theseZones: List[GEO] =
-          params.zoneQuadTree
-            .getRing(params.destinationUTM.getX, params.destinationUTM.getY, thisInnerRadius, thisOuterRadius)
-            .asScala
-            .toList
+          if (alternatives.isEmpty) {
+            _search(searchMode, parkingZoneIdsSeen, parkingZoneIdsSampled, iterations + 1)
```

Comment:
@dimaopen or @haitamlaarabi can fill you in on this one.

Reply:
It will stop if max search ring or ellipse is not infinite. I have renamed the function
`lookupParkingZones`
to
`lookupParkingZonesInNextSearchAreaUnlessThresholdReached`

Reply:
I see the maxsearch now. I'm never a fan of mutable recursive code, but I understand the need for it :)